tests/ClientResolverTest.php

```diff
@@ -1519,4 +1520,191 @@ public function testResolveFromEnvIniUseAwsSharedFiles(): void
             }
         }
     }
+
+    /**
+     * @param bool $isExpected
+     * @param array $expectedValue
+     * @param array $args
+     * @param array|null $fns
+     *
+     * @dataProvider resolvesAuthSchemePreferenceProvider
+     *
+     * @return void
+     */
+    public function testResolvesAuthSchemePreference(
+        bool $isExpected,
+        array $expectedValue,
+        array $args = [],
+        ?array $fns = null
+    ): void
+    {
+        try {
+            // To create the source such as env or from ini
+            if ($fns !== null && is_callable($fns['build_source'])) {
+                call_user_func($fns['build_source']);
+            }
+
+            $clientResolver = new ClientResolver(
+                ClientResolver::getDefaultArguments()
+            );
+            $resolvedConfig = $clientResolver->resolve([
+                    'service' => 's3',
+                    'region' => 'us-east-1'
+                ] + $args, new HandlerList());
+
+            if ($isExpected) {
+                $this->assertArrayHasKey(
+                    'auth_scheme_preference',
+                    $resolvedConfig
+                );
+                $this->assertTrue(
+                    is_array($resolvedConfig['auth_scheme_preference'])
+                );
+                $this->assertEquals(
+                    $expectedValue,
+                    $resolvedConfig['auth_scheme_preference']
+                );
+            } else {
+                $this->assertArrayNotHasKey(
+                    'auth_scheme_preference',
+                    $resolvedConfig
+                );
+            }
+        } finally {
+            if ($fns !== null && is_callable($fns['on_completion'])) {
+                call_user_func($fns['on_completion']);
+            }
+        }
+    }
+
+    /**
+     * @return Generator
+     */
+    public function resolvesAuthSchemePreferenceProvider(): Generator
```

Comment:
Since we're doing a lot of the same things repeatedly with the `fns`, can we not just create separate arguments for env and ini values and conditionally add those? Similarly, for cleanup, can we do that conditionally in the body of the test? That would make this test a bit more readable